### PR TITLE
DM-15173: fix & merge the antlr4 integration branch into master

### PR DESCRIPTION
**Pull request by n8pease**

### CODE REVIEW

File: core/modules/ccontrol/A4UserQueryFactory.cc [hunk 1]
```diff
@@ -34,29 +37,48 @@
...
-
+#include "util/IterableFormatter.h"
 
 
 // antlr4 C++ runtime seems to require that we use namespace antlr4; trying to use it with classes
 // results in at least one error where the `tree` decl assumes that it's already operating in the
 // antlr4 namespace. I may be wrong about this though; it needs research (before merge to master).
 // If I'm right we should see about fixing the antlr4 cpp runtime & issuing a PR to them.
 using namespace antlr4;
+using namespace std;
```
Comment (iagaponenko) [inline core/modules/ccontrol/A4UserQueryFactory.cc]: Are we really supposed to do this? This may open a whole "can of worms".

Comment (iagaponenko) [reply core/modules/ccontrol/A4UserQueryFactory.cc]: The comment above also applies to:
using namespace antlr4;

File: core/modules/ccontrol/A4UserQueryFactory.cc [hunk 2]
```diff
@@ -34,29 +37,48 @@
...
 
 namespace {
 LOG_LOGGER _log = LOG_GET("lsst.qserv.ccontrol.A4UserQueryFactory");
+
+// For the current query, this returns a list where each pair contains a bit of the string from the query
+// and how antlr4 tokenized that bit of string. It is useful for debugging problems where antlr4 did not
+// parse a query as expected, in the case where the string was not tokenized as expected.
+vector<pair<string, string>> getTokenPairs(CommonTokenStream & tokens, QSMySqlLexer & lexer) {
+    vector<pair<string, string>> ret;
+    for (auto t : tokens.getTokens()) {
```
Comment (iagaponenko) [inline core/modules/ccontrol/A4UserQueryFactory.cc]: The right way to iterate over collections in C++11/14 is:
for (auto&& t: ...) { ... }
This method would also take into account qualifiers (such as const).

File: core/modules/ccontrol/A4UserQueryFactory.cc [hunk 3]
```diff
@@ -34,29 +37,48 @@
...
+        string name = lexer.getVocabulary().getSymbolicName(t->getType());
```
Comment (iagaponenko) [inline core/modules/ccontrol/A4UserQueryFactory.cc]: could this be replaced with a more efficient:
auto const& name = ...

(provided the called method returns a constant reference to a string).

And the rest of the loop body could be reduced to:

ret.push_back(
  make_pair(
    name.empty() ?
      lexer.getVocabulary().getLiteralName(t->getType()) :
      name,
    t->getText()));

The later is just a matter of personal taste :-)

File: core/modules/parser/QSMySqlListener.cc [hunk 4]
```diff
@@ -1224,6 +1224,8 @@ class BinaryComparasionPredicateAdapter :
             compPredicate->op = SqlSQL2Tokens::LESS_THAN_OP;
         } else if ("<>" == _comparison) {
             compPredicate->op = SqlSQL2Tokens::NOT_EQUALS_OP;
+        } else if ("!=" == _comparison) {
```
Comment (iagaponenko) [inline core/modules/parser/QSMySqlListener.cc]: Do you need to support '<=' and '>=' as well?

File: core/modules/parser/QSMySqlParser.g4 [hunk 5]
```diff
@@ -4,6 +4,14 @@ import MySqlParser;
 
 options { tokenVocab=QSMySqlLexer; }
 
+// adds `MINUS?` before REAL_LITERAL
+constant
```
Comment (iagaponenko) [inline core/modules/parser/QSMySqlParser.g4]: I don't see the storm used in the rest of the gramma. Who's using it?
Also, there is one use of 'constants' in the gramma. Are these ('constant' and 'constants') related?

File: core/modules/ccontrol/A4UserQueryFactory.cc (hunk 1 above)
Comment (n8pease) [reply core/modules/ccontrol/A4UserQueryFactory.cc]: in a .cc it's ok. verboten in .h files.

File: core/modules/parser/QSMySqlListener.cc (hunk 4 above)
Comment (n8pease) [reply core/modules/parser/QSMySqlListener.cc]: it hasn't come up yet. (there's no test for it, so right now: no)

File: core/modules/parser/QSMySqlParser.g4 (hunk 5 above)
Comment (n8pease) [reply core/modules/parser/QSMySqlParser.g4]: `constants` uses `constant`. `constants` is used plenty in MySqlParser.g4

Comment (n8pease) [reply core/modules/parser/QSMySqlParser.g4]: e.g. https://github.com/lsst/qserv/blob/master/core/modules/parser/MySqlParser.g4#L723

Comment (iagaponenko) [reply core/modules/parser/QSMySqlParser.g4]: I see it! For some reason GitHub didn't show me the whole file, so I just missed those.

File: core/modules/ccontrol/A4UserQueryFactory.cc (hunk 1 above)
Comment (iagaponenko) [reply core/modules/ccontrol/A4UserQueryFactory.cc]: Ok, as long as you feel its safe 👍

File: core/modules/parser/QSMySqlParser.g4 (hunk 5 above)
Comment (n8pease) [reply core/modules/parser/QSMySqlParser.g4]: it's a different file :-)
overrides for that file are in this file.

Comment (iagaponenko) [reply core/modules/parser/QSMySqlParser.g4]: Thanks!

File: core/modules/ccontrol/A4UserQueryFactory.cc [hunk 6]
```diff
@@ -34,29 +37,48 @@
...
 
 namespace {
 LOG_LOGGER _log = LOG_GET("lsst.qserv.ccontrol.A4UserQueryFactory");
+
+// For the current query, this returns a list where each pair contains a bit of the string from the query
+// and how antlr4 tokenized that bit of string. It is useful for debugging problems where antlr4 did not
+// parse a query as expected, in the case where the string was not tokenized as expected.
+vector<pair<string, string>> getTokenPairs(CommonTokenStream & tokens, QSMySqlLexer & lexer) {
+    vector<pair<string, string>> ret;
+    for (auto&& t : tokens.getTokens()) {
```
Comment (n8pease) [inline core/modules/ccontrol/A4UserQueryFactory.cc]: fixed :-)

**Comment by n8pease** [comment]
this is merged to master & my local branch is in sync with the remove branch. I don't know why git didn't auto close it.

**Pull request closed without merging**